rust/lance-index/src/scalar/bitmap.rs

```diff
@@ -323,45 +339,95 @@ fn get_batch_from_arrays(
 }
 
-// Takes an iterator of Vec<u64> and processes each vector
-// to turn it into a RoaringTreemap. Each RoaringTreeMap is
-// serialized to bytes. The entire collection is converted to a BinaryArray
-fn get_bitmaps_from_iter<I>(iter: I) -> Arc<dyn Array>
-where
-    I: Iterator<Item = RowIdTreeMap>,
-{
-    let mut builder = BinaryBuilder::new();
-    iter.for_each(|bitmap| {
+fn chunked_keys_and_bitmaps(
+    state: HashMap<ScalarValue, RowIdTreeMap>,
+    value_type: &DataType,
+) -> Vec<(Arc<dyn Array>, Arc<dyn Array>)> {
+    let mut batches = Vec::new();
+    let mut cur_keys = Vec::new();
+    let mut cur_bitmaps = Vec::new();
+    let mut cur_bytes = 0;
+
+    for (key, bitmap) in state.into_iter() {
         let mut bytes = Vec::new();
         bitmap.serialize_into(&mut bytes).unwrap();
-        builder.append_value(&bytes);
-    });
+        let bitmap_len = bytes.len();
+
+        // If adding this bitmap would overflow, flush current batch
+        if cur_keys.len() >= MAX_BITMAP_ARRAY_LENGTH
+            || cur_bytes + bitmap_len > MAX_BITMAP_ARRAY_LENGTH
+        {
+            let keys_array = if cur_keys.is_empty() {
+                new_empty_array(value_type)
+            } else {
+                ScalarValue::iter_to_array(cur_keys.clone().into_iter()).unwrap()
+            };
+            let mut binary_builder = BinaryBuilder::new();
+            for b in &cur_bitmaps {
+                binary_builder.append_value(b);
+            }
+            let bitmaps_array = Arc::new(binary_builder.finish()) as Arc<dyn Array>;
+            batches.push((keys_array, bitmaps_array));
+            cur_keys.clear();
+            cur_bitmaps.clear();
+            cur_bytes = 0;
+        }
 
-    Arc::new(builder.finish())
+        cur_keys.push(key);
+        cur_bitmaps.push(bytes);
+        cur_bytes += bitmap_len;
+    }
+
+    // Flush any remaining
+    if !cur_keys.is_empty() {
+        let keys_array = ScalarValue::iter_to_array(cur_keys).unwrap();
+        let mut binary_builder = BinaryBuilder::new();
+        for b in &cur_bitmaps {
+            binary_builder.append_value(b);
+        }
+        let bitmaps_array = Arc::new(binary_builder.finish()) as Arc<dyn Array>;
+        batches.push((keys_array, bitmaps_array));
+    }
+
+    batches
 }
 
 async fn write_bitmap_index(
     state: HashMap<ScalarValue, RowIdTreeMap>,
     index_store: &dyn IndexStore,
     value_type: &DataType,
 ) -> Result<()> {
-    let keys_iter = state.keys().cloned();
-    let keys_array = if state.is_empty() {
-        new_empty_array(value_type)
+    let batches = chunked_keys_and_bitmaps(state, value_type);
+
+    // Create a schema even if we have no batches
+    let schema = if batches.is_empty() {
+        // For empty state, create a schema with the same structure
+        Arc::new(Schema::new(vec![
+            Field::new("keys", value_type.clone(), true),
+            Field::new("bitmaps", DataType::Binary, true),
+        ]))
```

Comment:
If we know the schema then why do we need this if / else branch?

Reply:
refactored